### PR TITLE
lint: add evalOrder checker

### LINKED ISSUE #73: lint: add "dangerous depencency on evaluation order" checker
Some expressions and statements in Go may evaluate differently on various implementations due to unspecified evaluation order details.

See https://github.com/Quasilyte/goism/issues/115 for more info.

### PR DESCRIPTION
**Pull request by Quasilyte**
Needed IsTypeExpr, so moved its implementation from
astwalk to a new package lintutil, so both lint and
astwalk can use it.

As 2 checkers now use AstSet-like type, moved it to lintutil.

Fixes #73

### CODE REVIEW

**Comment by Quasilyte** [comment]
Not very powerful and does sometimes report false positives (it does not verify that function that takes a pointer really modifies a value).
Experimental until this issue is resolved.

File: lint/internal/lintutil/ast_set.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package lintutil
```
Comment (cristaloleg) [inline lint/internal/lintutil/ast_set.go]: WDYT s/`ast_set.go`/`astset.go` ? or even `set.go`

File: lint/evalOrder_checker.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+func (c *evalOrderChecker) VisitStmt(stmt ast.Stmt) {
+	switch stmt := stmt.(type) {
+	case *ast.ReturnStmt:
+		c.checkReturn(stmt)
+	case *ast.AssignStmt:
+		// TODO(quasilyte):
+		//	https://github.com/golang/go/issues/23188
+		//	https://github.com/golang/go/issues/24448
+		//	https://github.com/golang/go/issues/23017
```
Comment (cristaloleg) [inline lint/evalOrder_checker.go]: + https://github.com/golang/go/issues/25609 ?

Comment (Quasilyte) [reply lint/evalOrder_checker.go]: It's referenced inside test files.

File: lint/internal/lintutil/ast_set.go (hunk 1 above)
Comment (Quasilyte) [reply lint/internal/lintutil/ast_set.go]: Fixed.

File: lint/evalOrder_checker.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		return
+	}
+	c.reset()
+	for _, x := range ret.Results {
+		c.collectDeps(x)
+	}
+	// TODO(quasilyte): we can parametrize the threshold later.
+	if deps := c.depsCount(); deps != 0 {
+		c.warn(ret, deps)
+		return
```
Comment (Quasilyte) [inline lint/evalOrder_checker.go]: This return is redundant.

**Pull request merged**